File: Framework/PythonInterface/plugins/algorithms/FindEPP.py
# pylint: disable=no-name-in-module
from __future__ import (absolute_import, division, print_function)
from mantid.api import PythonAlgorithm, AlgorithmFactory, MatrixWorkspaceProperty, ITableWorkspaceProperty
from mantid.kernel import Direction
from mantid.simpleapi import CreateEmptyTableWorkspace, DeleteWorkspace, Fit
import numpy as np


class FindEPP(PythonAlgorithm):
    def __init__(self):
        """
        Init
        """
        PythonAlgorithm.__init__(self)
        self.workspace = None

    def category(self):
        return "Workflow\\MLZ\\TOFTOF;Utility"

    def name(self):
        return "FindEPP"

    def summary(self):
        return "Performs Gaussian fit of all spectra to find the elastic peak position."

    def PyInit(self):
        # input
        self.declareProperty(MatrixWorkspaceProperty("InputWorkspace", "", direction=Direction.Input),
                             doc="Input Sample or Vanadium workspace")

        # output
        self.declareProperty(ITableWorkspaceProperty("OutputWorkspace", "", Direction.Output),
                             doc="The name of the table workspace that will be created.")

    def do_fit_gaussian(self, index):
        """
        Performs gaussian fit of the specified spectrum
        returns fitStatus, chiSq, covarianceTable, paramTable
        """
        result = None

        x_values = np.array(self.workspace.readX(index))
        y_values = np.array(self.workspace.readY(index))

        # get peak centre position, assuming that it is the point with the highest value
        imax   = np.argmax(y_values)
        height = y_values[imax]

        # check for zero or negative signal
        if height <= 0.:
            self.log().warning("Workspace %s, detector %d has maximum <= 0" % (self.workspace.name(), index))
            return result

        # Get the positions around the max, where the y value first drops down 0.5*height
        # This is a better approach when there is a higher noise in the tails in real data
        fwhm_right = x_values.size - 1 - imax # number of bins from max to the right end
        fwhm_left = imax                      # number of bins from max to the left end

        if imax < y_values.size - 1:
            y_right = y_values[imax+1:]
            y_right_below = np.argwhere(y_right < 0.5 * height)
            if y_right_below.size != 0:
                fwhm_right = y_right_below[0][0]

        if imax > 0:
            y_left = y_values[:imax]
            y_left = y_left[::-1]
            y_left_below = np.argwhere(y_left < 0.5 * height)
            if y_left_below.size != 0:
                fwhm_left = y_left_below[0][0]

        fwhm_pos = np.minimum(imax+fwhm_right,x_values.size-1)
        fwhm_neg = np.maximum(imax-fwhm_left,0)

        self.log().debug("(spectrum %d) : FWHM lower edge is %d, upper edge is %d" % (index,fwhm_neg,fwhm_pos))

        if fwhm_pos - fwhm_neg + 1 < 3:
            self.log().warning("Spectrum " + str(index) + " in workspace " + self.workspace.name() +
                               " has a too narrow peak. Cannot guess sigma. Check your data.")
            return result

        fwhm = x_values[fwhm_pos] - x_values[fwhm_neg]

        self.log().debug("(spectrum %d) : FWHM is %f" % (index, fwhm))

        sigma = fwhm / (2.*np.sqrt(2.*np.log(2.)))

        # execute Fit algorithm
        tryCentre = x_values[imax]
        fitFun = "name=Gaussian,PeakCentre=%s,Height=%s,Sigma=%s" % (tryCentre,height,sigma)

        startX = tryCentre - 3.0*fwhm
        endX   = tryCentre + 3.0*fwhm

        tempOutputPrefix = "__EPPfit_" + str(self.workspace) + "_" + str(index)
        # pylint: disable=assignment-from-none
        # result = fitStatus, chiSq, covarianceTable, paramTable
        result = Fit(InputWorkspace=self.workspace, WorkspaceIndex=index, StartX = startX, EndX=endX,
                     Output=tempOutputPrefix, Function=fitFun, CreateOutput=True, OutputParametersOnly=True)
        return result

    def PyExec(self):
        self.workspace = self.getProperty("InputWorkspace").value
        outws_name = self.getPropertyValue("OutputWorkspace")

        # create table and columns
        outws = CreateEmptyTableWorkspace(OutputWorkspace=outws_name)
        columns = ["PeakCentre", "PeakCentreError", "Sigma", "SigmaError", "Height", "HeightError", "chiSq"]
        nextrow = dict.fromkeys(["WorkspaceIndex"] + columns + ["FitStatus"])
        outws.addColumn(type="int", name="WorkspaceIndex", plottype=1)  # x
        for col in columns:
            outws.addColumn(type="double", name=col)
        outws.addColumn(type="str", name="FitStatus")

        nb_hist = self.workspace.getNumberHistograms()
        for idx in range(nb_hist):
            nextrow["WorkspaceIndex"] = idx
            result = self.do_fit_gaussian(idx)
            if not result:
                for col in columns:
                    nextrow[col] = 0
                nextrow["FitStatus"] = "failed"
            else:
                nextrow["FitStatus"] = result[0]
                nextrow["chiSq"] = result[1]
                ptable = result[3]
                for num in range(ptable.rowCount() - 1):
                    row = ptable.row(num)
                    name = row["Name"]
                    nextrow[name] = row["Value"]
                    nextrow[name+"Error"] = row["Error"]
                DeleteWorkspace(result.OutputParameters)
                DeleteWorkspace(result.OutputNormalisedCovarianceMatrix)
            outws.addRow(nextrow)
        self.setProperty("OutputWorkspace", outws)
        return

AlgorithmFactory.subscribe(FindEPP)

File: Framework/Algorithms/src/FindEPP.cpp
// Mantid Repository : https://github.com/mantidproject/mantid
//
// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
//     NScD Oak Ridge National Laboratory, European Spallation Source
//     & Institut Laue - Langevin
// SPDX - License - Identifier: GPL - 3.0 +
#include "MantidAlgorithms/FindEPP.h"
#include "MantidAPI/TableRow.h"
#include "MantidDataObjects/TableWorkspace.h"


#include <cmath>
#include <sstream>

namespace Mantid {
namespace Algorithms {

using namespace Mantid::Kernel;
using namespace Mantid::API;
using namespace Mantid::DataObjects;

// Register the algorithm into the AlgorithmFactory
DECLARE_ALGORITHM(FindEPP)

//----------------------------------------------------------------------------------------------

/// Algorithms name for identification. @see Algorithm::name
const std::string FindEPP::name() const { return "FindEPP"; }

/// Algorithm's version for identification. @see Algorithm::version
int FindEPP::version() const { return 2; }

/// Algorithm's category for identification. @see Algorithm::category
const std::string FindEPP::category() const {
  return "Workflow\\MLZ\\TOFTOF;Utility";
}

/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
const std::string FindEPP::summary() const {
  return "Performs Gaussian fits over each spectrum to find the Elastic Peak "
         "Position (EPP).";
}

//----------------------------------------------------------------------------------------------
/** Initialize the algorithm's properties.
 */
void FindEPP::init() {
  declareProperty(std::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
                      "InputWorkspace", "", Direction::Input),
                  "An input workspace.");
  declareProperty(std::make_unique<WorkspaceProperty<API::ITableWorkspace>>(
                      "OutputWorkspace", "", Direction::Output),
                  "An output workspace.");
}

//----------------------------------------------------------------------------------------------
/** Execute the algorithm.
 */
void FindEPP::exec() {
  m_inWS = getProperty("InputWorkspace");

  initWorkspace();

  int64_t numberspectra = static_cast<int64_t>(m_inWS->getNumberHistograms());

  // Loop over spectra
  PARALLEL_FOR_IF(threadSafe(*m_inWS, *m_outWS))
  for (int64_t index = 0; index < numberspectra; ++index) {
    PARALLEL_START_INTERUPT_REGION
    fitGaussian(index);
    PARALLEL_END_INTERUPT_REGION
  }
  PARALLEL_CHECK_INTERUPT_REGION

  setProperty("OutputWorkspace", m_outWS);
}

/* Call Fit as child algorithm for each spectra
 * @param index : the workspace index
 */
void FindEPP::fitGaussian(int64_t index) {
  size_t spectrum = static_cast<size_t>(index);
  m_outWS->cell<int>(spectrum, 0) = static_cast<int>(spectrum);

  const auto &x = m_inWS->x(spectrum).rawData();
  const auto &y = m_inWS->y(spectrum).rawData();
  const auto &e = m_inWS->e(spectrum).rawData();

  // Find the maximum value and it's index
  const auto maxIt = std::max_element(y.begin(), y.end());
  const double height = *maxIt;
  size_t maxIndex = static_cast<size_t>(std::distance(y.begin(), maxIt));

  if (height > 0) {
    // Find how many bins are around maximum, that are above half-maximum
    // Initialize the distances of the half-maxima bins from maximum
    size_t leftHalf = maxIndex, rightHalf = x.size() - maxIndex - 1;

    // Find the first bin on the right side of maximum, that drops below
    // half-maximum
    for (auto it = maxIt; it != y.end(); ++it) {
      if (*it < 0.5 * height) {
        rightHalf = it - maxIt - 1;
        break;
      }
    }

    // Find the first bin on the left side of maximum, that drops below
    // half-maximum
    for (auto it = maxIt; it != y.begin(); --it) {
      if (*it < 0.5 * height) {
        leftHalf = maxIt - it - 1;
        break;
      }
    }
    g_log.debug() << "Peak in spectrum #" << spectrum
                  << " has last bins above 0.5*max at " << leftHalf << "\t"
                  << rightHalf << "\n";

    // We want to fit only if there are at least 3 bins (including the maximum
    // itself) above half-maximum
    if (rightHalf + leftHalf >= 2) {

      // Prepare the initial parameters for the fit
      double fwhm = x[maxIndex + rightHalf] - x[maxIndex - leftHalf];
      double sigma = fwhm / (2. * sqrt(2. * log(2.)));
      double center = x[maxIndex];
      double start = center - 3. * fwhm;
      double end = center + 3. * fwhm;

      std::stringstream function;
      function << "name=Gaussian,PeakCentre=";
      function << center << ",Height=" << height << ",Sigma=" << sigma;

      g_log.debug() << "Fitting spectrum #" << spectrum
                    << " with: " << function.str() << "\n";

      IAlgorithm_sptr fitAlg = createChildAlgorithm("Fit", 0., 0., false);
      fitAlg->setProperty("Function", function.str());
      fitAlg->setProperty("InputWorkspace", m_inWS);
      fitAlg->setProperty("WorkspaceIndex", static_cast<int>(spectrum));
      fitAlg->setProperty("StartX", start);
      fitAlg->setProperty("EndX", end);
      fitAlg->setProperty("CreateOutput", true);
      fitAlg->setProperty("OutputParametersOnly", true);
      fitAlg->executeAsChildAlg();

      const std::string status = fitAlg->getProperty("OutputStatus");
      ITableWorkspace_sptr fitResult = fitAlg->getProperty("OutputParameters");

      if (status == "success") {
        m_outWS->cell<double>(spectrum, 1) = fitResult->cell<double>(1, 1);
        m_outWS->cell<double>(spectrum, 2) = fitResult->cell<double>(1, 2);
        m_outWS->cell<double>(spectrum, 3) = fitResult->cell<double>(2, 1);
        m_outWS->cell<double>(spectrum, 4) = fitResult->cell<double>(2, 2);
        m_outWS->cell<double>(spectrum, 5) = fitResult->cell<double>(0, 1);
        m_outWS->cell<double>(spectrum, 6) = fitResult->cell<double>(0, 2);
        m_outWS->cell<double>(spectrum, 7) = fitResult->cell<double>(3, 1);
        m_outWS->cell<std::string>(spectrum, 8) = status;
      } else {
        g_log.debug() << "Fit failed in spectrum #" << spectrum
                      << ". \nReason :" << status
                      << ". \nSetting the maximum.\n";
        m_outWS->cell<std::string>(spectrum, 8) = "fitFailed";
        m_outWS->cell<double>(spectrum, 1) = x[maxIndex];
        m_outWS->cell<double>(spectrum, 2) = 0.;
        m_outWS->cell<double>(spectrum, 5) = height;
        m_outWS->cell<double>(spectrum, 6) = e[maxIndex];
      }

    } else {
      g_log.information() << "Found <=3 bins above half maximum in spectrum #"
                          << index << ". Not fitting.\n";
      m_outWS->cell<std::string>(spectrum, 8) = "narrowPeak";
      m_outWS->cell<double>(spectrum, 1) = x[maxIndex];
      m_outWS->cell<double>(spectrum, 2) = 0.;
      m_outWS->cell<double>(spectrum, 5) = height;
      m_outWS->cell<double>(spectrum, 6) = e[maxIndex];
    }
  } else {
    g_log.notice() << "Negative maximum in spectrum #" << spectrum
                   << ". Skipping.\n";
    m_outWS->cell<std::string>(spectrum, 8) = "negativeMaximum";
  }
  m_progress->report();
}

/**
 * Initializes the output workspace
 */
void FindEPP::initWorkspace() {

  m_outWS = boost::make_shared<TableWorkspace>();

  const std::vector<std::string> columns = {
      "PeakCentre", "PeakCentreError", "Sigma", "SigmaError",
      "Height",     "HeightError",     "chiSq"};

  m_outWS->addColumn("int", "WorkspaceIndex");
  m_outWS->getColumn(0)->setPlotType(1);
  for (const auto &column : columns) {
    m_outWS->addColumn("double", column);
  }
  m_outWS->addColumn("str", "FitStatus");

  const size_t numberSpectra = m_inWS->getNumberHistograms();
  m_progress = std::make_unique<Progress>(this, 0.0, 1.0, numberSpectra);

  m_outWS->setRowCount(numberSpectra);
}

} // namespace Algorithms
} // namespace Mantid
